`NIOSSLHandler`: behave sensibly on `close(mode: .output)`

felixschlegel opened the pull request:
### Motivation

If NIOSSLHandler receives `close(mode: .output) or close(mode: .input)`, it ignores them, optionally failing their promises. This is weird behaviour. It should do better.

### Modifications

* flush outstanding messages on `NIOSSLHandler.close(mode: .output)`
* close SSL Connection on `NIOSSLHandler.close(mode: .output)`
* support closing the input of `NIOSSLHandler` when used as a server with `ChannelOptions.Types.AllowRemoteHalfClosureOption` enabled
* trigger TCP close when NIOSSLHandler.state == .inputClose
* write integration tests to verify the behaviour

> **Warning**: This PR relies on `EmbeddedChannel` supporting `ChannelOptions.Types.AllowRemoteHalfClosureOption`, which is currently not possible and [proposed as a Patch to `swift-nio`](https://github.com/apple/swift-nio/pull/2429)

> **Note**: I have removed a lot of whitespace, so consider selecting `Hide Whitespace` in the GitHub `Files Changed` tab

felixschlegel commented on Sources/NIOSSL/NIOSSLHandler.swift:
```diff
@@ -205,14 +210,52 @@ public class NIOSSLHandler : ChannelInboundHandler, ChannelOutboundHandler, Remo
+        case .idle, .closed, .unwrapping, .unwrapped:
+            // For idle, closed, unwrapping, and unwrapped connections we immediately pass this on to the next
+            // channel handler.
+            context.close(mode: .output, promise: promise)
```
Question: what do you think of this behaviour? Passing on `context.close(mode: .output)` can result in errors if the next `ChannelHandler` does not support `CloseMode.output`?

Lukasa replied:
Passing this on is fine.

Lukasa commented on Sources/NIOSSL/NIOSSLHandler.swift:
```diff
@@ -102,14 +107,18 @@ public class NIOSSLHandler : ChannelInboundHandler, ChannelOutboundHandler, Remo
     }
 
     public func channelActive(context: ChannelHandlerContext) {
+        context.channel.getOption(ChannelOptions.allowRemoteHalfClosure).whenSuccess { halfClosureAllowed in
```
I don't think caching this value is semantically quite right: users can change this value during the lifetime of the `Channel`. We should look it up when we need it.

Similarly, we should try to use the sync options where we can.

Lukasa commented on Sources/NIOSSL/NIOSSLHandler.swift:
```diff
@@ -190,7 +199,53 @@ public class NIOSSLHandler : ChannelInboundHandler, ChannelOutboundHandler, Remo
     }
-    
+
+    // TODO: how do I test this?
```
With `EmbeddedChannel`, you can fire the input closed event into the pipeline manually.

Lukasa commented on Sources/NIOSSL/NIOSSLHandler.swift:
```diff
@@ -190,7 +199,53 @@ public class NIOSSLHandler : ChannelInboundHandler, ChannelOutboundHandler, Remo
+            userInboundInputClosedTriggered(context: context)
+        default:
+            context.fireUserInboundEventTriggered(event)
+            // Trigger closure of TCP input as this event has not occured although we are in state .inputClosed already.
```
This code seems to be in the wrong place.

Lukasa commented on Sources/NIOSSL/NIOSSLHandler.swift:
```diff
@@ -190,7 +199,53 @@ public class NIOSSLHandler : ChannelInboundHandler, ChannelOutboundHandler, Remo
+        let channelError: NIOSSLError
+        switch self.state {
+        case .closed, .idle:
+            return
```
In both of these cases we should probably still send the user event onwards.

Lukasa commented on Sources/NIOSSL/NIOSSLHandler.swift:
```diff
@@ -190,7 +199,53 @@ public class NIOSSLHandler : ChannelInboundHandler, ChannelOutboundHandler, Remo
+        case .closed, .idle:
+            return
+        case .inputClosed:
+            context.fireUserInboundEventTriggered(ChannelEvent.inputClosed)
```
Ideally this isn't right. We should synthesise this event when we _enter_ the `.inputClosed` state, and then suppress it here.

Lukasa commented on Sources/NIOSSL/NIOSSLHandler.swift:
```diff
@@ -200,20 +255,70 @@ public class NIOSSLHandler : ChannelInboundHandler, ChannelOutboundHandler, Remo
             // we should not flush immediately as we have not completed the handshake and instead buffer the flush
             self.bufferFlush()
-        case .active, .unwrapping, .closing, .unwrapped, .closed:
+        case .active, .unwrapping, .closingOutput, .closing, .unwrapped, .inputClosed, .outputClosed, .closed: // TODO: .closingOutput + .inputClosed correct here?
```
All of these are correct.

Lukasa commented on Sources/NIOSSL/NIOSSLHandler.swift:
```diff
@@ -200,20 +255,70 @@ public class NIOSSLHandler : ChannelInboundHandler, ChannelOutboundHandler, Remo
+            self.closeOutput(context: context, promise: promise)
+        case .all:
+            self.closeAll(context: context, promise: promise)
+        default:
```
May as well write `case .input` as it's the only case left and take advantage of the exhaustive switch.

Lukasa commented on Sources/NIOSSL/NIOSSLHandler.swift:
```diff
@@ -200,20 +255,70 @@ public class NIOSSLHandler : ChannelInboundHandler, ChannelOutboundHandler, Remo
+            // We flush all outstanding writes once the handshake step is complete and set our state to .outputClosed aftwerwards.
+            // This prevents any further writes to this channel.
+            if let promise = promise, let closeOutputPromise = self.closeOutputPromise {
+                closeOutputPromise.futureResult.cascade(to: promise)
```
Can we possibly have a `closeOutputPromise` if we're not in state `closingOutput`?

Lukasa commented on Sources/NIOSSL/NIOSSLHandler.swift:
```diff
@@ -200,20 +255,70 @@ public class NIOSSLHandler : ChannelInboundHandler, ChannelOutboundHandler, Remo
+            } else if let promise = promise {
+                self.closeOutputPromise = promise
+                self.state = .closingOutput(self.scheduleTimedOutShutdown(context: context))
+                self.doShutdownStep(context: context)
```
It seems wrong to only do the shutdown if there's a promise.

Lukasa commented on Sources/NIOSSL/NIOSSLHandler.swift:
```diff
@@ -200,20 +255,70 @@ public class NIOSSLHandler : ChannelInboundHandler, ChannelOutboundHandler, Remo
+                self.doShutdownStep(context: context)
+            }
+
+            self.flush(context: context)
```
I think we have to flush _before_ we start the shutdown.

Lukasa commented on Sources/NIOSSL/NIOSSLHandler.swift:
```diff
@@ -200,20 +255,70 @@ public class NIOSSLHandler : ChannelInboundHandler, ChannelOutboundHandler, Remo
+
+            self.flush(context: context)
+
+            self.closeOutputPromise?.futureResult.whenSuccess {
```
This is incorrect: if we don't have a promise, this never fires.

This makes me thing that maybe `outputClosing` isn't necessary at all. It's a sort of immediate transition: we send our CLOSE_NOTIFY and the process is conceptually "done". So maybe this is an atomic transition.

felixschlegel replied:
Ok, I have added two new commits, one removing the `.outputClosing` state and one removing the `closeOutputPromise` — let me know what you think!

Lukasa commented on Sources/NIOSSL/NIOSSLHandler.swift:
```diff
@@ -391,6 +503,8 @@ public class NIOSSLHandler : ChannelInboundHandler, ChannelOutboundHandler, Remo
 
             switch targetCompleteState {
+            case .outputClosed:
+                break
```
I think this is a relevant edge: if it happens, we end up immediately in full closed.

Lukasa commented on Sources/NIOSSL/NIOSSLHandler.swift:
```diff
@@ -469,15 +553,24 @@ public class NIOSSLHandler : ChannelInboundHandler, ChannelOutboundHandler, Remo
-                    self.state = .closing(self.scheduleTimedOutShutdown(context: context))
-                case .unwrapping, .closing:
+                case .active, .outputClosed:
+                    if self.getAllowRemoteHalfClosureFromChannel(context: context) == false {
```
Ideally we'd cache this across the call to `doFlushReadData`.

felixschlegel replied:
Good point!

Lukasa commented on Sources/NIOSSL/NIOSSLHandler.swift:
```diff
@@ -491,6 +584,16 @@ public class NIOSSLHandler : ChannelInboundHandler, ChannelOutboundHandler, Remo
+        var halfClosureAllowed = false
+        if let syncOptions = context.channel.syncOptions {
+            let result = try? syncOptions.getOption(ChannelOptions.allowRemoteHalfClosure)
+            result.map { halfClosureAllowed = $0 }
```
Using `map` for the side-effect is a bit weird here, it's probably better to use `if let`.

Lukasa commented on Sources/NIOSSL/NIOSSLHandler.swift:
```diff
@@ -722,7 +825,33 @@ extension NIOSSLHandler {
 
+    /// Completes the given promise when all outstanding writes have been flushed.
+    private func bufferedWritesCascadeFlushPromise(
+        to promise: EventLoopPromise<Void>?,
```
I don't think this use of the promise is correct. It should be attached to the `close(mode: .output)` operation. We should _fail_ it if we don't actually issue that call, but we shouldn't succeed it ourselves.

Lukasa commented on Sources/NIOSSL/NIOSSLHandler.swift:
```diff
@@ -190,7 +192,48 @@ public class NIOSSLHandler : ChannelInboundHandler, ChannelOutboundHandler, Remo
+            // these writes.
+            channelError = NIOSSLError.uncleanShutdown
+        }
+        context.fireErrorCaught(channelError)
```
Can we reverse the order of these two events? The error is more prominent, I think, especially when we hit a ragged EOF.

Lukasa commented on Sources/NIOSSL/NIOSSLHandler.swift:
```diff
@@ -200,19 +244,56 @@ public class NIOSSLHandler : ChannelInboundHandler, ChannelOutboundHandler, Remo
+            flushCompletePromise.futureResult.whenComplete { _ in
+                self.doShutdownStep(context: context)
+            }
+            bufferedWritesCascadeFlushPromise(to: flushCompletePromise, context: context)
```
So this promise dance is probably unnecessary. We can send `closeOutput` once we've called `flush`, even if the write promises haven't completed. Handlers are required to queue `close(mode: .output)` behind writes.

Lukasa commented on Sources/NIOSSL/NIOSSLHandler.swift:
```diff
@@ -342,7 +423,7 @@ public class NIOSSLHandler : ChannelInboundHandler, ChannelOutboundHandler, Remo
                 completeHandshake(context: context)
             }
-        case .unwrapping, .closing, .unwrapped, .closed:
+        case .unwrapping, .closing, .unwrapped, .closed, .inputClosed, .outputClosed:
```
I think this is wrong, `.inputClosed` and `.outputClosed` should match `active` here I think.

felixschlegel commented:
Hey @Lukasa ,

I have implemented your requested changes and added the buffering of `closeOutput` when we are still in `state = .handshaking / .additionalVerification`.

I still have a couple of open questions for this PR:

* should we also support `scheduledShutdown` for `close(mode: .output)` (see the normal close mode for reference)
* what if the downstream `ChannelHandler` does not support `close(mode: .output)`? At the moment we just invoke `context.close(mode: .output)` hoping for the next `ChannelHandler` not to fail

Best regards,
-Felix

Lukasa commented:
@swift-server-bot add to allowlist

Lukasa commented:
> should we also support scheduledShutdown for close(mode: .output) (see the normal close mode for reference)

No, scheduled shutdown is waiting for us to get a `CLOSE_NOTIFY` from the remote peer. In this mode, we don't expect one.

> what if the downstream ChannelHandler does not support close(mode: .output)?

Then the partial close fails and the user should upgrade to full close.

Lukasa commented on Sources/NIOSSL/NIOSSLHandler.swift:
```diff
@@ -732,27 +865,29 @@ extension NIOSSLHandler {
-            let bufferedWrite = self.bufferedWrites.removeFirst()
-            bufferedWrite.promise?.fail(reason)
+    private func discardBufferedActions(reason: Error) {
+        while self.bufferedActions.count > 0 {
```
This can be more concisely expressed as `while let bufferedAction = self.bufferedActions.popFirst()`.

Lukasa commented on Sources/NIOSSL/NIOSSLHandler.swift:
```diff
@@ -778,13 +920,28 @@ extension NIOSSLHandler {
         } catch {
             // We encountered an error, it's cleanup time. Close ourselves down.
-            channelClose(context: context, reason: error)
+            if case .outputClosed = self.state {} else {
```
This is a really weird construction, prefer using a `switch`.

felixschlegel commented:
@swift-server-bot test this please

yim-lee commented:
@swift-server-bot test this please

Lukasa commented on Sources/NIOSSL/NIOSSLHandler.swift:
```diff
@@ -391,6 +484,10 @@ public class NIOSSLHandler : ChannelInboundHandler, ChannelOutboundHandler, Remo
             switch targetCompleteState {
+            case .outputClosed:
+                // No full channel close here. We would expect users to invoke a full close regardless of
+                // previously completed half closures.
```
This comment is out of date now, right?

felixschlegel replied:
Thanks for pointing that out! I have updated the comment so that it reflects the current state of our half-closure behaviour 😄

Lukasa commented on Sources/NIOSSL/NIOSSLHandler.swift:
```diff
@@ -778,13 +933,33 @@ extension NIOSSLHandler {
-            channelClose(context: context, reason: error)
+            switch self.state {
+            case .outputClosed:
+                break
```
Why do we suppress error handling in this state?

felixschlegel replied:
Fixed 🔨
